SI-9105 Fix EnclosingMethod for classes defined in lambdas

lrytz opened the pull request:
The reason for an inconsistent EnclosingMethod attribute was the
symbol owner chain. Initially, closure class symbols don't exist, they
are only created in UnCurry (delambdafy:inline). So walking the
originalOwner of a definition does not give yield closure classes.

The commit also fixes uses of isAnonymousClass, isAnonymousFunction
and isDelambdafyFunction in two ways:
1. by phase-travelling to an early phase. after flatten, the name
    includes the name of outer classes, so the properties may become
    accidentally true (they check for a substring in the name)
2. by ensuring that the (destructive) renames during LambdaLift
    don't make the above properties accidentally true. This was in
    fact the cause for SI-8900.

lrytz committed:
SI-9105 Fix EnclosingMethod for classes defined in lambdas

The reason for an inconsistent EnclosingMethod attribute was the
symbol owner chain. Initially, closure class symbols don't exist, they
are only created in UnCurry (delambdafy:inline). So walking the
originalOwner of a definition does not give yield closure classes.

The commit also fixes uses of isAnonymousClass, isAnonymousFunction
and isDelambdafyFunction in two ways:

  1. by phase-travelling to an early phase. after flatten, the name
     includes the name of outer classes, so the properties may become
     accidentally true (they check for a substring in the name)

  2. by ensuring that the (destructive) renames during LambdaLift
     don't make the above properties accidentally true. This was in
     fact the cause for SI-8900.

lrytz commented:
[`-Ybackend:GenBCode -Ydelambdafy:method` build](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1377/) passed, and also a [community build with `-Ybackend:GenASM -Ydelambdafy:inline`](https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/123/).

lrytz commented:
Review by @adriaanm

adriaanm commented on src/compiler/scala/tools/nsc/transform/LambdaLift.scala:
```diff
@@ -256,15 +256,33 @@ abstract class LambdaLift extends InfoTransform {
+         * Ensures that name mangling does not accidentally make a class respond `true` to any of
+         * isAnonymousClass, isAnonymousFunction, isDelambdafyFunction, e.g. by introducing "$anon".
+         */
+        def ensureNonAnon(prefix: String) = {
```
While this seems like a good idea, I worry about doing this in 2.11.x.

adriaanm commented on src/compiler/scala/tools/nsc/transform/LambdaLift.scala:
```diff
@@ -256,15 +256,33 @@ abstract class LambdaLift extends InfoTransform {
+          val anoncls = nme.ANON_CLASS_NAME.toString
+          val anonfun = nme.ANON_FUN_NAME.toString
+          val delamda = nme.DELAMBDAFY_LAMBDA_CLASS_NAME.toString
+          prefix.replace(anoncls, nestedIn + anoncls.replace("$", ""))
```
I'd introduce constants for `nestedIn + anoncls.replace("$", "")` and friends.

retronym commented:
Let's add the test case submitted yesterday in [SI-9124](https://issues.scala-lang.org/browse/SI-9124).

retronym commented:
We should also run the following check over all the JARs we produce to find if we spot any ill-formed class metadata.

``` scala
import scala.reflect.io._

object Test {
  def main(args: Array[String]): Unit = {
    val jarFile = AbstractFile.getDirectory(new java.io.File(args(0)))
    def flatten(f: AbstractFile): Iterator[AbstractFile] =
      if (f.isClassContainer) f.iterator.flatMap(flatten)
      else Iterator(f)

    val classFQNs = flatten(jarFile).filter(_.hasExtension("class")).map(_.path.replaceAllLiterally("/", ".").replaceAll(".class", ""))
    val classes = classFQNs.map(name => Class.forName(name))
    val okay = classes.map(cls => (cls, try {Left(cls.getEnclosingMethod) } catch { case t: Throwable => Right(t) })).filter(_._2.isRight)
    println(okay.mkString("\n"))  
  }
}
```

```
(export V=2.11.4; ~/scala/$V/bin/scalac sandbox/test.scala && ~/scala/$V/bin/scala Test ~/scala/$V/lib/scala-library.jar)
(class scala.reflect.ClassManifestDeprecatedApis$$anonfun$subargs$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$iterateUntilEmpty$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$filterImpl$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.immutable.MapProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.immutable.SetProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqViewLike$Reversed$$anonfun$createReversedIterator$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqLike$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqLike$$anonfun$occCounts$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.GenMapLike$$anonfun$liftedTree1$1$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.parallel.ParMapLike$$anon$4,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.parallel.ParMapLike$$anon$3,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$3,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$scala$collection$TraversableLike$$sliceInternal$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.mutable.MapProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.mutable.FlatHashTable$$anonfun$checkConsistent$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$iterateUntilEmpty$2,Right(java.lang.InternalError: Enclosing method not found))
```

lrytz commented:
thanks for the test, @retronym. It passes after my patch, but I extended your test a bit and found more issues. I'll look into this. Here's the current version:

``` scala
import scala.reflect.io._

object Test {
  def main(args: Array[String]): Unit = {
    val jarFile = AbstractFile.getDirectory(new java.io.File(args(0)))
    def flatten(f: AbstractFile): Iterator[AbstractFile] =
      if (f.isClassContainer) f.iterator.flatMap(flatten)
      else Iterator(f)

    val classFQNs = flatten(jarFile).filter(_.hasExtension("class")).map(_.path.replace("/", ".").replaceAll(".class$", ""))
    val classes = classFQNs.map(name => Class.forName(name))
    val okay = classes.map(cls => (cls, 
      try { Left(List(cls.getEnclosingMethod, cls.getEnclosingClass, cls.getEnclosingConstructor, cls.getDeclaredClasses)) }
      catch { case t: Throwable => Right(t) })).filter(_._2.isRight)
    println(okay.mkString("\n"))
  }
}
```

It gives a number of messages like this one:

```
(class scala.sys.process.ProcessBuilder$Sink$class,Right(java.lang.IncompatibleClassChangeError: scala.sys.process.ProcessBuilder and scala.sys.process.ProcessBuilder$Sink$class disagree on InnerClasses attribute))
```

lrytz committed:
Fix InnerClass / EnclosingMethod for impl and specialized classes

Trait implementation classes and specialized classes are always
considered top-level in terms of the InnerClass / EnclosingMethod
attributes. These attributes describe source-level properties, and
such classes are a compilation artifact.

Note that the same is true for delambdafy:method closure classes
(they are always top-level).

lrytz committed:
SI-9124 fix EnclosingMethod of classes nested in implOnly trait defs

Private trait methods are not added to the generated interface, they
end up only in the implementation class. For classes nested in such
methods, the EnclosingMethod attribute was incorrect.

Since the EnclosingMethod attribute expresses a source-level property,
but the actual enclosing method does not exist in the bytecode, we
set the enclosing method to null.

lrytz commented:
almost there.. still have two issues

```
(class scala.tools.nsc.interpreter.JavapClass$fun$u0020with$u0020files$$anonfun$scala$tools$nsc$interpreter$JavapClass$fun$u0020with$u0020files$$lookup$1$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.tools.nsc.interpreter.JavapClass$ClassLoaderOps$$anonfun$alldirs$1$1,Right(java.lang.InternalError: Enclosing method not found))
```

lrytz commented:
these are due to value classes, i'll check tomorrow. the rest should be ready for review by @adriaanm (i addressed your feedback from the last review).

adriaanm commented:
ok!

som-snytt commented:
Some guys really overdo it with the value classes. They're so good, you can't have just one.

retronym commented:
I have often wondered if it would have been better to put the value class method bodies in static methods in the value class, rather than in instance methods on the companion. Same story for defaults in constructor. 

Admittedly the status of STATIC methods is a bit grey within the compilation pipeline, but leaving code in its natural habitat rather than shunting it to the companion might avoid a few headaches.

lrytz commented:
PLS REBUILD/pr-scala@071ea9716e45a7af080c5ea6e85ced38e08d52d6

scala-jenkins commented:
(kitty-note-to-self: ignore 72622164)
:cat: Roger! Rebuilding pr-scala for 071ea971. :rotating_light:

lrytz commented:
[-Ybackend:GenBCode -Ydelambdafy:method build](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1379/console) OK

lrytz committed:
Fix InnerClass / EnclosingMethod for closures nested in value classes

Members of value classes are moved over to the companion object early.
This change ensures that closure classes nested in value classes
appear that way to Java reflection.

lrytz committed:
Test java reflection on scala library / reflect / compiler jars

Run a number of Java reflection operations on all classes in
scala library / reflect / compiler.

The test is based on a draft by Jason.

lrytz commented:
fixed the value classes and added @retronym's test

retronym commented:
We've picked up an inliner warning en route:

```
% diff /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/presentation/t7678-presentation.log /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/presentation/t7678.check
@@ -1,2 +1 @@
-warning: there were 5 inliner warnings; re-run with -Yinline-warnings for details
 reload: TypeTag.scala
```

adriaanm commented on src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:
```diff
@@ -595,7 +595,8 @@ abstract class GenASM extends SubComponent with BytecodeWriters { self =>
           val x = innerClassSymbolFor(s)
           if(x ne NoSymbol) {
             assert(x.isClass, "not an inner-class symbol")
-            val isInner = !x.rawowner.isPackageClass
```
Is this the only changed file under `src/compiler/scala/tools/nsc/backend/jvm/` that's relevant to the official/default compilation pipeline?

lrytz replied:
No, the changes in [BCodeAsmCommon](https://github.com/scala/scala/pull/4264/files#diff-1) also apply to GenASM

adriaanm commented on src/interactive/scala/tools/nsc/interactive/Global.scala:
```diff
@@ -152,7 +152,7 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
 
   // don't keep the original owner in presentation compiler runs
   // (the map will grow indefinitely, and the only use case is the backend)
-  override protected def saveOriginalOwner(sym: Symbol) { }
```
can this be problematic for the IDE/Ensime/...?

lrytz replied:
good point, i'll go back to the original api

lrytz commented:
PLS REBUILD/pr-scala@5ee04deffe699d1d7cd0ec54c05cb692beff2abd

lrytz commented:
PLS REBUILD/pr-scala@5ee04deffe699d1d7cd0ec54c05cb692beff2abd

lrytz commented:
PLS REBUILD/pr-scala@9d90035bc67c1a05d3831fd0c15ab032c3637b9f

lrytz commented:
bye bye kitty

scala-jenkins commented:
Bye bye Lukas.

lrytz commented:
:yum:

gkossakowski commented:
:grin:
